subsys/rpmsg_service/rpmsg_service.c

```diff
@@ -0,0 +1,162 @@
+ */
+
+#include <rpmsg_service/rpmsg_service.h>
+
+#include "rpmsg_backend.h"
+
+#include <zephyr.h>
+#include <device.h>
+#include <logging/log.h>
+
+#include <openamp/open_amp.h>
+
+#define LOG_LEVEL LOG_LEVEL_INFO
+#define LOG_MODULE_NAME rpmsg_service
+LOG_MODULE_REGISTER(LOG_MODULE_NAME);
+
+/* Configuration defines */
+#if !DT_HAS_CHOSEN(zephyr_ipc_shm)
+#error "Module requires definition of shared memory for rpmsg"
+#endif
+
+#define MASTER IS_ENABLED(CONFIG_RPMSG_SERVICE_MODE_MASTER)
+
+static struct virtio_device vdev;
+static struct rpmsg_virtio_device rvdev;
+static struct metal_io_region *io;
+
+#if MASTER
+static struct rpmsg_virtio_shm_pool shpool;
+#endif
+
+static struct {
+	const char *name;
+	rpmsg_ept_cb cb;
+	struct rpmsg_endpoint ep;
+	volatile bool bound;
+} endpoints[CONFIG_RPMSG_SERVICE_NUM_ENDPOINTS];
+
+static void rpmsg_service_unbind(struct rpmsg_endpoint *ep)
+{
+	rpmsg_destroy_ept(ep);
+}
+
+#if MASTER
+
+static void ns_bind_cb(struct rpmsg_device *rdev,
+					const char *name,
+					uint32_t dest)
+{
+	int err;
+
+	for (int i = 0; i < CONFIG_RPMSG_SERVICE_NUM_ENDPOINTS; ++i) {
+		if (strcmp(name, endpoints[i].name) == 0) {
+			err = rpmsg_create_ept(&endpoints[i].ep,
+						   rdev,
+						   name,
+						   RPMSG_ADDR_ANY,
+						   dest,
+						   endpoints[i].cb,
+						   rpmsg_service_unbind);
+
+			if (err != 0) {
+				LOG_ERR("Creating remote endpoint %s"
+					" failed wirh error %d", name, err);
+			} else {
+				endpoints[i].bound = true;
+			}
+
+			return;
+		}
+	}
+
+	LOG_ERR("Remote endpoint %s not registered locally", name);
+}
+
+#endif
+
+static int rpmsg_service_init(const struct device *dev)
+{
+	int32_t err;
+
+	(void)dev;
+
+	LOG_DBG("RPMsg service initialization start");
+
+	err = rpmsg_backend_init(&io, &vdev);
+	if (err) {
+		LOG_ERR("RPMsg backend init failed with error %d", err);
+		return err;
+	}
+
+#if MASTER
+	rpmsg_virtio_init_shm_pool(&shpool, (void *)SHM_START_ADDR, SHM_SIZE);
+	err = rpmsg_init_vdev(&rvdev, &vdev, ns_bind_cb, io, &shpool);
+#else
+	err = rpmsg_init_vdev(&rvdev, &vdev, NULL, io, NULL);
+#endif
+
+	if (err) {
+		LOG_ERR("rpmsg_init_vdev failed %d", err);
+		return err;
+	}
+
+#if !MASTER
+	struct rpmsg_device *rdev;
+
+	rdev = rpmsg_virtio_get_rpmsg_device(&rvdev);
+
+	for (int i = 0; i < CONFIG_RPMSG_SERVICE_NUM_ENDPOINTS; ++i) {
+		if (endpoints[i].name) {
+			err = rpmsg_create_ept(&endpoints[i].ep,
+						rdev,
+						endpoints[i].name,
+						RPMSG_ADDR_ANY,
+						RPMSG_ADDR_ANY,
+						endpoints[i].cb,
+						rpmsg_service_unbind);
+
+			if (err) {
+				LOG_ERR("rpmsg_create_ept failed %d", err);
+				return err;
+			}
+		}
+	}
+#endif
+
+	LOG_DBG("RPMsg service initialized");
+
+	return 0;
+}
+
+int rpmsg_service_register_endpoint(const char *name, rpmsg_ept_cb cb)
+{
+	for (int i = 0; i < CONFIG_RPMSG_SERVICE_NUM_ENDPOINTS; ++i) {
+		if (!endpoints[i].name) {
+			endpoints[i].name = name;
+			endpoints[i].cb = cb;
+
+			return i;
+		}
+	}
```

Comment:
This imposes that every endpoints have to been registered before calling rpmsg_service_init. Any reason? Could become quite complex to resolve init priorities.
A simple test could be added here to test if the rpmsg is initialized and then rpmsg_create_ept could be called to dynamically register the new end point.

Reply:
I did my first attempt in a similar way you suggest. However, there was a problem with races between master and slave nodes (endpoint already registered in one node, but not yet in another one). To resolve race conditions, I've added the limitation to register endpoints before the service is initialized. All use-cases I have and samples tested work correctly with this limitation.

I think that we can work at robust way of removing this limitation later.

Reply:
Ok to treat limitation later but I think this function should at least return  an error if called after the init step for the slave.